src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -26,47 +24,63 @@ internal partial class SQLitePersistentStorage
+        private abstract class Accessor<TKey, TDatabaseKey>
+            where TDatabaseKey : struct
         {
             protected readonly SQLitePersistentStorage Storage;
+            protected readonly Table Table;
+
+            private readonly ImmutableArray<(string name, string type)> _primaryKeyColumns;
+            private readonly ImmutableArray<(string name, string type)> _allColumns;
 
-            // Cache the statement strings we want to execute per accessor.  This way we avoid
-            // allocating these strings each time we execute a command.  We also cache the prepared
-            // statements (at the connection level) we make for each of these strings.  That way we
-            // only incur the parsing cost once. After that, we can use the same prepared statements
-            // and just bind the appropriate values it needs into it.
+            // Cache the statement strings we want to execute per accessor.  This way we avoid allocating these strings
+            // each time we execute a command.  We also cache the prepared statements (at the connection level) we make
+            // for each of these strings.  That way we only incur the parsing cost once. After that, we can use the same
+            // prepared statements and just bind the appropriate values it needs into it.
             //
-            // values like 0, 1, 2 in the name are the `?`s in the sql string that will need to be
+            // Names starting with numbers (like 0primarykey) indicates the `?`s in the sql string that will need to be
             // bound to runtime values appropriately when executed.
 
-            private readonly string _select_rowid_from_main_table_where_0;
-            private readonly string _select_rowid_from_writecache_table_where_0;
-            private readonly string _insert_or_replace_into_writecache_table_values_0_1_2;
             private readonly string _delete_from_writecache_table;
             private readonly string _insert_or_replace_into_main_table_select_star_from_writecache_table;
-
-            public Accessor(SQLitePersistentStorage storage)
+            private readonly string _select_rowid_from_main_table_where_0primarykey;
+            private readonly string _select_rowid_from_writecache_table_where_0primarykey;
+            private readonly string _insert_or_replace_into_writecache_table_values_0primarykey_1checksum_2data;
+
+            public Accessor(
+                Table table,
+                SQLitePersistentStorage storage,
+                params (string name, string type)[] primaryKeysArray)
             {
-                var main = Database.Main.GetName();
+                Table = table;
+                Storage = storage;
+
+                _primaryKeyColumns = primaryKeysArray.ToImmutableArray().Add((DataNameIdColumnName, SQLiteIntegerType));
+                _allColumns = _primaryKeyColumns.Add((ChecksumColumnName, SQLiteBlobType)).Add((DataColumnName, SQLiteBlobType));
```

Comment:
does _allColumns need to include primary key columsn (for example DataNameIdColumnName?

Reply:
yes, all columsn includes all primary key columns (including the data-name column).  that's what hte above logic accomplishes.

Reply:
ah oops I misread what the starting array was